src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = GetWordList();
+                    string tag = lhs == "category" ? "cat" : lhs;
+                    switch (op)
+                    {
+                        case "=":
+                        case "==":
+                            return string.Format("<{0}>{1}</{0}>", tag, values);
+                        case "!=":
+                            return string.Format("<not><{0}>{1}</{0}></not>", tag, values);
+                        default:
+                            return string.Format("<{0} op='{1}'>{2}</{0}>", tag, op, values);
+                    }
+
+                case "name":
+                case "test":
+                case "fullname":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = Peek().Kind == TokenKind.String
+                        ? GetString()
+                        : GetWord();
+                    tag = lhs == "fullname" ? "test" : lhs;
```

Comment:
That's a good point. Two names should mean two things. With cat and category, it's obvious that we have an abbreviation, so no sweat. But test and fullname are a bit confusing.

Here is a list of my lhs keywords and what they mean:
   id = the secret id of the test, known to NUnit and not of much use except for a gui.
   name = the Name field of the test, which may have parentheses and arguments, etc.(name in the XML)
   test (and fullname) = the FullName field of the test, just as we now use it for the --test option (fullname in XML)
   method = the Name field of the MethodInfo for the test (methodName in XML)
   class = the FullName field of the class (className in XML)
The reason class and method is what they are is so that a runner can find the source.

What would be a good substitute for test? Would you change any of the others?